clang/lib/Driver/Driver.cpp

```diff
@@ -4995,14 +4995,33 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
 
   // For an FPGA archive, we add the unbundling step above to take care of
   // the device side, but also unbundle here to extract the host side
-  for (const auto &LI : LinkerInputs) {
+  for (auto &LI : LinkerInputs) {
     Action *UnbundlerInput = nullptr;
+    auto wrapObject = [&] {
+      bool EarlyLink = false;
+      if (const Arg *A = Args.getLastArg(options::OPT_fsycl_link_EQ))
+        EarlyLink = A->getValue() == StringRef("early");
+      if (EarlyLink && Args.hasArg(options::OPT_fintelfpga)) {
+        // Only wrap the object with -fsycl-link=early
+        auto *BC = C.MakeAction<OffloadWrapperJobAction>(LI, types::TY_LLVM_BC);
+        auto *ASM = C.MakeAction<BackendJobAction>(BC, types::TY_PP_Asm);
+        LI = C.MakeAction<AssembleJobAction>(ASM, types::TY_Object);
+      }
+    };
     if (auto *IA = dyn_cast<InputAction>(LI)) {
       if (IA->getType() == types::TY_FPGA_AOCR ||
           IA->getType() == types::TY_FPGA_AOCX) {
         // Add to unbundler.
         UnbundlerInput = LI;
+      } else {
+        std::string FileName = IA->getInputArg().getAsString(Args);
+        if ((IA->getType() == types::TY_Object && !isObjectFile(FileName)) ||
+            IA->getInputArg().getOption().hasFlag(options::LinkerInput))
+          continue;
+        wrapObject();
       }
+    } else {
+      wrapObject();
```

Comment:
Which of the user scenarios is this block aimed at? (for personal clarification)

Reply:
The `LinkerInputs` here are not considered `InputAction` (i.e. they didn't come directly from the user) but are derived from source files.  We need to wrap those objects as well before creating the early image.

Reply:
Right. The only point that remains for me, what's going to happen if any of the `LinkerInputs` turns out to be a _fat_ object? Is this an imaginable scenario?

Reply:
@mdtoguchi, could you please comment on that one?

Reply:
Oops - I must have read it and not responded.  If any of the `LinkerInputs` are device libraries or fat objects, they will not be in this else section - those are considered `InputAction` from the user and will be separately processed.